libs/rotary-encoder/rotary.cpp

```diff
@@ -0,0 +1,81 @@
+#include "pxt.h"
+
+// update sim if modifying these
+#define ROT_EV_TIMER 0x1233
+#define ROT_EV_CHANGED 0x2233
+
+const static int8_t posMap[] = {0, +1, -1, +2, -1, 0, -2, +1, +1, -2, 0, -1, +2, -1, +1, 0};
+
+class RotaryEncoder_ {
+  public:
+    uint16_t id;
+    uint16_t state;
+    int position;
+    Pin &pinA, &pinB;
+
+    void process(Event) {
+        // based on comments in https://github.com/PaulStoffregen/Encoder/blob/master/Encoder.h
+        uint16_t s = state & 3;
+        if (pinA.getDigitalValue())
+            s |= 4;
+        if (pinB.getDigitalValue())
+            s |= 8;
+
+        state = (s >> 2);
+        if (posMap[s]) {
+            position += posMap[s];
+            Event ev(id, ROT_EV_CHANGED);
+        }
+    }
+
+    RotaryEncoder_(Pin &pinA, Pin &pinB) : pinA(pinA), pinB(pinB) {
+        position = 0;
+        id = pinA.id;
+        // don't do exactly 1000us, so that it doesn't occur exactly at scheduler ticks
+        system_timer_event_every_us(973, id, ROT_EV_TIMER);
+        EventModel::defaultEventBus->listen(id, ROT_EV_TIMER, this, &RotaryEncoder_::process,
+                                            MESSAGE_BUS_LISTENER_IMMEDIATE);
+    }
+};
+
+typedef class RotaryEncoder_ *RotaryEncoder;
+
+/**
+ * Rotary and other encoders
+ */
+namespace encoders {
+/**
+ * Create a new rotary encoder connected to given pins
+ */
+//% blockId=inputCreateRotaryEncoder block="create rotary encoder with pins $pinA and B $pinB"
+//% weight=99
```

Comment:
can you use this ctor function in blocks if you have `fixedInstances` on it?

Reply:
no you can't so i'm not exposing it in blocks.